solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,109 @@
+     */
+    public static boolean isCompressedBytes(byte[] data) {
+        if (data == null || data.length < 2) return false;
+        // two characters form the hex value.
+        char[] hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[0] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[0] & 0xF), 16);
+        if (!Arrays.equals(hexDigits, ZLIB_FIRST_MAGIC_HEX)) return false;
+        hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[1] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[1] & 0xF), 16);
+        return Arrays.equals(hexDigits, ZLIB_SECOND_MAGIC_HEX);
+    }
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data, 0, data.length);
+        // Attempt to get the decompressed size from trailing bytes, this will be present if compressed by Solr
+        int decompressedSize = ByteBuffer.wrap(data, data.length - 8, 4).getInt();
+        int xoredSize = ByteBuffer.wrap(data, data.length - 4, 4).getInt();
+        if ((decompressedSize ^ COMPRESSED_SIZE_MAGIC_NUMBER) != xoredSize) {
+            // Take best guess of decompressed size since it wasn't included in trailing bytes
+            decompressedSize = (int)(1.8*data.length);
+        }
+        byte[] buf = new byte[decompressedSize];
+        int actualDecompressedSize = 0;
+        while (!inflater.finished()) {
+            if (actualDecompressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
+            actualDecompressedSize += inflater.inflate(buf, actualDecompressedSize, decompressedSize);
+        }
+        return Arrays.copyOf(buf, actualDecompressedSize);
+    }
+
+    /**
+     * Compresses bytes into zlib compressed bytes using java native compression
+     * @param data the input uncompressed data to be compressed
+     * @return zlib compressed bytes
+     * @throws IOException
+     */
+    public static byte[] compressBytes(byte[] data) throws IOException {
+        Deflater compressor = new Deflater(Deflater.BEST_SPEED);
+        compressor.setInput(data);
+        compressor.finish();
+        byte[] buf = new byte[data.length+8];
+        int compressedSize = 0;
+        while (!compressor.finished()) {
+            if (compressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
+            compressedSize += compressor.deflate(buf, compressedSize, buf.length - compressedSize);
+        }
+        if (buf.length != compressedSize + 8) {
```

Comment:
in contrast to the decompress case, where we very often expect to have an appropriately sized buffer that we can simply return, here for the compress case it's _very_ unlikely the size will be exactly right (to the extent that it's probably not worth checking this case, and resize the array unconditionally?).